ethcore/verification/src/queue/mod.rs

```diff
@@ -470,8 +470,10 @@ impl<K: Kind, C> VerificationQueue<K, C> {
 	pub fn import(&self, input: K::Input) -> Result<H256, (K::Input, Error)> {
 		let hash = input.hash();
 		let raw_hash = input.raw_hash();
+		let mut processing = self.processing.write();
```

Comment:
this will essentially make importing to the queue single threaded
let me run an import bench to see how it affects the perf

Reply:
Ah yeah good point

Reply:
> this will essentially make importing to the queue single threaded

Is this accurate? I probably misunderstand something but isn't the job of `import()` to:
1. run checks to ensure we're not redoing work we needn't/mustn't do 
2. instantiate an unverified Block or Header "item"
3. enqueue the item and notify verifier threads that there's new work to do

So the critical part wrt to parallelism comes after the call to `import()`? And is it really critical that we can **enqueue** new work in parallel? It would surprise me if this was the bottleneck.

That said, talking to @tomusdrw, he makes a good argument for not bothering with fixing this potential race because a) the scenario is unlikely to occur due to how sync works (highly ordered) and b) the worst case scenario is (I believe) that we end up doing the same verification work twice (two verification threads racing to verify the item), which is not really that bad.

If the reasoning above is correct then we can land this PR and make enqueuing new items essentially single threaded but there's little value to be gained by doing so: a toss-up.

Reply:
Some initial verification is (has to be) done already in `K::create` afair, so it's not as simple as enqueing the item for later heavy verification.

I believe that similar to previous PR, we are doing a trade-off here of possibly doing the cheap verification twice in favour of lock-free checks.
Imagine if I spam the node with invalid headers now, and all networking threads will be just waiting for that lock to learn that the blocks are invalid. So we never insert anything, and just spend a lot of time blocking other tasks by holding `write` lock.

Reply:
> Some initial verification is (has to be) done already in `K::create` afair

[`verify_block_basic()`](https://github.com/paritytech/parity-ethereum/blob/5bd6b208af94efe87c20a1931c3e7896543c402e/ethcore/verification/src/queue/kind.rs#L95) or [`verify_header_params()`](https://github.com/paritytech/parity-ethereum/blob/5bd6b208af94efe87c20a1931c3e7896543c402e/ethcore/verification/src/queue/kind.rs#L186) according to what `K::Item` is. 
If either of those calls are too expensive we're in trouble in the case of a spam storm, lock or no lock right?

Reply:
Well, it should be cheap to verify that, so it's fine to call it in a blocking fashion from multiple threads, but not good if we make it sequential, that's my point.

Reply:
By cheap verification, you mean phase 1 verification right?

However, if `K::create` returns Ok for the same block twice (seems possible) it will fool the `total difficulty` calculation and enqueue the block twice in `unverified`, no?

It wouldn't hurt to document because it was clearly not obvious to me :P

Reply:
So the issue only arises when submitting same blocks.
We can see if the return value here
https://github.com/paritytech/parity-ethereum/blob/a9214081c0aca86720e61ac3a132d711a28d075f/ethcore/verification/src/queue/mod.rs#L493

is `Some` and in that case return `Ok(hash)`. Would that solve the incorrect total difficulty problem, @niklasad1 ?

Reply:
actually we can return `return Err((input, Error::Import(ImportError::AlreadyQueued).into()));` in that case

Reply:
yeah, sounds good to me

Reply:
I would suggest we change the error type of `import` as well.
Currently it returns a pair as an error `Result<_, (K::Input, Error)>` and the input is only seems to be used for `EthcoreError::Block`, maybe the input should be part of it, or would it create an unwanted dependency?

Reply:
> or would it create an unwanted dependency?

Yeah I think it'd be problematic unfortunately.

Reply:
hmm, I don't think it is possible to return `return Err((input, Error::Import(ImportError::AlreadyQueued).into()));` without modifying the `Kind` trait after `Kind::create` has been called.

Because `Kind::create` takes ownership over `input` and when it is `Ok` returns `K::Unverified`.
In the implementation,  `K::Input` and `K::Unverified` are the same type but the generic function can't know that. This is quite neat to have type safety for the transitions but in order to work, we would need to clone `Unverified` which is quite expensive (Header + body),

Thus, having for example `EthcoreError::Block((Option<T>, BlockError))` would be quite neat